core/src/components/alert/alert.tsx

```diff
@@ -462,12 +462,12 @@ export class Alert implements ComponentInterface, OverlayInterface {
   private async buttonClick(button: AlertButton) {
     const role = button.role;
     const values = this.getValues();
-    if (isCancel(role)) {
+    if (isCancel(role) && button.handler === undefined) {
```

Comment:
I'll need to ask the team, but this change does change the current behavior of a `cancel` role button with a custom handler that returns `false`. 

Previously, it would have dismissed the overlay. With the proposed changes it would not dismiss. 

I _think_ the change in behavior is valid/expected, but never hurts to get a second opinion 🙂

Reply:
Hello! After speaking with the team, I'd advise we adjust the implementation pattern here. We should de-risk and validate a use case for needing to prevent a dismiss with a cancel role, before making that change. 

The proposed fix for the original issue should execute the handler callback, but not prevent dismissing. Something to the effect of:
```ts
private async buttonClick(button: AlertButton) {
  const role = button.role;
  const values = this.getValues();
  const returnData = await this.callButtonHandler(button, values);
  if (isCancel(role)) {
    /**
     * Cancel roles should always dismiss the alert,
     * even if the handler function returns `false`.
     */
    return this.dismiss({ values, ...returnData }, role);
  }
  if (returnData !== false) {
    return this.dismiss({ values, ...returnData }, role);
  }
  return false;
}
```

While investigating this issue further, I found that [`ion-action-sheet`](https://github.com/ionic-team/ionic-framework/blob/51c729eafc3b290a5daddf7f0ccffd0a2a9fe2aa/core/src/components/action-sheet/action-sheet.tsx#L252-L262), [`ion-picker`](https://github.com/ionic-team/ionic-framework/blob/51c729eafc3b290a5daddf7f0ccffd0a2a9fe2aa/core/src/components/picker/picker.tsx#L296-L306), and [`ion-toast`](https://github.com/ionic-team/ionic-framework/blob/51c729eafc3b290a5daddf7f0ccffd0a2a9fe2aa/core/src/components/toast/toast.tsx#L523-L533) all have very similar implementations and likely the same problematic behavior. 

We can discuss more in Slack, but I think addressing the fix across all these implementations would be great to knock out together, once we validate the path for one. I can help out here if desired.